tools/cldr-code/src/main/java/org/unicode/cldr/util/XPathValue.java

```diff
@@ -1,14 +1,33 @@
 package org.unicode.cldr.util;
 
+import java.util.Map;
+
 /** This is an interface for a read-only XPath. */
 public interface XPathValue {
+    /** How many elements are in this xpath? */
+    int size();
 
     /** Get the nth element. Negative values are from end */
     String getElement(int i);
 
+    /** See if the xpath contains an element */
+    public boolean containsElement(String element);
+
+    /**
+     * Get the attributes for the nth element (negative index is from end). Returns null or an empty
+     * map if there's nothing. PROBLEM: exposes internal map
```

Comment:
It would be better to return an ImmutableMap; that way we are guaranteed that nothing can change it.

Can discuss tradeoffs offline.

Reply:
That is, something like the following.

- from public Element makeImmutable() {
- extract the following into a method in a common location        
```
            if (attributes != null && !(attributes instanceof ImmutableMap)) {
                attributes = ImmutableMap.copyOf(attributes);
           }
```
- From public XPathParts freeze() {
- extract the following into a method in a common location, changing to call the above method instead of Element.makeImmutable()
```
            // ensure that it can't be modified. Later we can fix all the call sites to check
            // frozen.
            List<Element> temp = new ArrayList<>(elements.size());
            for (Element element : elements) {
                temp.add(element.makeImmutable());
            }
            elements = ImmutableList.copyOf(temp);
```
- then change the following to call that method after the handleParse call
```
    SimpleXPathParts(String xpath) {
        handleParse(xpath, true);
    }
```

There are alternatives, eg copy the code into SimpleXPathParts, or have an Elements base class, etc.

Reply:
`PROBLEM: exposes internal map` comment and code is from the original XPathParts implementation.  Perhaps that's a good follow up item as a separate concern from this one?

Reply:
Makes sense. Can you clone a ticket for that?

Reply:
https://unicode-org.atlassian.net/browse/CLDR-17565